### PR TITLE
ggml : change ggml_graph_compute() API to not require context

### LINKED ISSUE #287: ggml : `ggml_graph_compute` should not require `ggml_context`
We now have the following signature:

```c
    GGML_API void ggml_graph_compute(struct ggml_context * ctx, struct ggml_cgraph * cgraph);
```

The provided context is used during the computation to potentially allocate a work buffer needed by some of the `ggml` operators. Not only the buffer is not always needed, but having to pass an entire context is a poor design choice.

We need to avoid this by allowing the user to pass a work buffer (pointer and size) and an API that allows the user to query what is the work size needed to evaluate a specific compute graph. This way, the user will first make the query and then will be responsible to provide the work buffer externally if necessary

### PR DESCRIPTION
**Pull request by mqy**
Try resolve https://github.com/ggerganov/ggml/issues/287

EDIT: see latest update at the end.

<del>
### Intro

The design is a bit different to the suggested one: named the buffer type as  a generalized one:`ggml_cgraph_context`.

```
struct ggml_cgraph_context {
    size_t work_size;
    void * work_data;
    bool   planned; // true means ready to compute graph nodes.
};
```
I'll explain `planned` later, let's focus on the APIs.

The first parameter `ctx` of `ggml_graph_compute()` is deprecated (pass in NULL is OK).
Removed `wsize` and `work` from `ggml_cgraph`, unlikely break external users because no reason to use them directly.

To avoid break external users, can not simply change the signature of `ggml_graph_compute()`, have to add `ggml_graph_compute_v2()`, the name looks weird but this is the reality :/  Now `ggml_graph_compute()` is a thin wrapper of `ggml_graph_compute_v2().

Extracted codes to new function `ggml_graph_compute_plan()`: set node->`n_tasks`, calculate `work_size`.

### Usage
```
struct ggml_cgraph_context ctx = {0};

// case 1:
ggml_graph_compute_plan(&ctx, cgraph);
if (ctx.work_size > 0) {
    ctx.work_data = malloc(ctx.work_size);
}
ggml_graph_compute_v2(&ctx, cgraph);

// case 2:
ggml_graph_compute_v2(&ctx, cgraph);

// case 3:
ggml_graph_compute_v2(NULL, cgraph);

// case 4:
ggml_graph_compute(ggml_context *ctx, cgraph);

// case 5:
ggml_graph_compute(NULL, cgraph);
```

### Why did I add the field `planned`?

Because:
 * The `ctx` is allowed to be NULL, empty or initialized by `ggml_graph_compute_plan()`.
 * One of the corner cases is: the `work_size` and `work_data` can be default values even if the plan has been called.  So we can not simply determine whether or not call `ggml_graph_compute_plan()` by default values.
 * `ggml_graph_compute_plan()` MUST be called because it  also sets `node->n_tasks`. The `work_size` depends on `n_tasks`.

The `planned` makes plan-compute sequence stateful, not good enough. Any ideas?

</del>

# Update on JUL 3

1. No longer consider backward compatibility, this means the `plan phase` MUST be executed before `compute`. And `gml_graph_compute()` no longer responsible for creating any kind of buffer. @ggerganov
2. Removed `n_tasks` from `ggml_tensor`; removed `n_threads` from `ggml_cgraph`. @slaren
3. the `struct ggml_graph_compute_plan` implies that it should be initialized or created by some procedure. @howard0su

Usage:

```
struct ggml_graph_compute_plan plan = ggml_graph_compute_make_plan(gb, params.n_threads);
if (plan.work_size > 0) {
    plan.work_data = malloc(plan.work_size);
    GGML_ASSERT(plan.work_data);
}
ggml_graph_compute(&plan, gb);
if (plan.work_data) {
    free(plan.work_data);
}
```

I tested `main` and `perplexity`.

# Update JUL 6

### Summary
🛠️📚🚀

This pull request improves the performance and usability of the `ggml_graph_compute` function by introducing a new `ggml_cplan` structure and a helper function `ggml_graph_compute_helper`. It also adds support for specifying the number of command buffers for the Metal context, and updates the examples, tests, and workflows to use the new API and settings.

> _This pull request makes some changes_
> _To `ggml_graph_compute` and friends_
> _It adds a new plan_
> _And some Metal commands_
> _And fixes some warnings and ends_

### Walkthrough
*  Change the `ggml_graph_compute` function to require a `ggml_cplan` argument and add new functions `ggml_graph_plan` and `ggml_graph_compute_with_ctx` to support the new API ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-f0f2d0dc971e0aa60560e7e3bc1d512b4bf914aedf44333f7008c605433cd394L1300-R1314))
*  Add a helper function `ggml_graph_compute_helper` to wrap the logic of creating and using a `ggml_cplan` structure and update the example code in `ggml.h` to use the new API ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-256be3a51fd8a6c1332bbcf2d5dd122810ba137a6184495fc3e42415857f69baR34-R44), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-6a296d1a1222746faa37384705c8ae3f535428097dbe46295d7ccf8405a6ff07R23-R33), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267R63-R73), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-f0f2d0dc971e0aa60560e7e3bc1d512b4bf914aedf44333f7008c605433cd394L68-R68), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR82-R100))
*  Remove the fields `n_threads`, `n_tasks`, and `work` from the `ggml_cgraph` and `ggml_tensor` structures and add them to the `ggml_cplan` structure ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267L1429-R1442), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-f0f2d0dc971e0aa60560e7e3bc1d512b4bf914aedf44333f7008c605433cd394L421-L423), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-f0f2d0dc971e0aa60560e7e3bc1d512b4bf914aedf44333f7008c605433cd394L435-R453))
*  Update the `llama_eval_internal` function to use the new API and adjust the number of threads for the BLAS settings ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL1268-R1289), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL1309-R1334), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL1594-R1623), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL1614-R1646))
*  Add a parameter `n_cb` to the `ggml_metal_init` function and a function `ggml_metal_set_n_cb` to control the number of command buffers for the Metal context and update the `ggml_metal_graph_compute` function to use the `n_cb` field instead of the `n_threads` field ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-3cebf7a1bc643557eff27cae1c81d9894615a7f944c271eefcfca861aaa013e1L37-R43), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d189c5117255f368aa89f521bd523afa399ada0bbfb7043b74c65bf4ac2472fdR28-R29), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d189c5117255f368aa89f521bd523afa399ada0bbfb7043b74c65bf4ac2472fdL89-R96), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d189c5117255f368aa89f521bd523afa399ada0bbfb7043b74c65bf4ac2472fdR214-R217), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d189c5117255f368aa89f521bd523afa399ada0bbfb7043b74c65bf4ac2472fdL357-R364))
*  Update the `metal.cpp` example to pass the number of threads to the `ggml_metal_init` function ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-f5177339d387bef8794ff76aa0cdae0aa28b68d63067bfc2ffe82904c096b619L38-R40))
*  Update the `baby-llama.cpp`, `benchmark-matmult.cpp`, and `train-text-from-scratch.cpp` examples to use the new API and remove the assignments of `gf.n_threads` ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-256be3a51fd8a6c1332bbcf2d5dd122810ba137a6184495fc3e42415857f69baR1583-R1584), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-256be3a51fd8a6c1332bbcf2d5dd122810ba137a6184495fc3e42415857f69baL1589), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-256be3a51fd8a6c1332bbcf2d5dd122810ba137a6184495fc3e42415857f69baL1598-R1610), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-256be3a51fd8a6c1332bbcf2d5dd122810ba137a6184495fc3e42415857f69baL1614-R1626), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-256be3a51fd8a6c1332bbcf2d5dd122810ba137a6184495fc3e42415857f69baL1662), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-256be3a51fd8a6c1332bbcf2d5dd122810ba137a6184495fc3e42415857f69baL1668-R1679), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-256be3a51fd8a6c1332bbcf2d5dd122810ba137a6184495fc3e42415857f69baL1690-R1706), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-6a296d1a1222746faa37384705c8ae3f535428097dbe46295d7ccf8405a6ff07L162-R181), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-6a296d1a1222746faa37384705c8ae3f535428097dbe46295d7ccf8405a6ff07L190), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-6a296d1a1222746faa37384705c8ae3f535428097dbe46295d7ccf8405a6ff07L202-R213), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-6a296d1a1222746faa37384705c8ae3f535428097dbe46295d7ccf8405a6ff07L224-R235), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-6a296d1a1222746faa37384705c8ae3f535428097dbe46295d7ccf8405a6ff07L231-R242), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-6a296d1a1222746faa37384705c8ae3f535428097dbe46295d7ccf8405a6ff07L256-R267), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267R3174), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267R3196-R3197), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267L3220-L3222), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267L3251-R3260), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267L3275-R3284), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267L3357), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267L3363-R3371), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-d1491dab0ced83c9f5cd0ea092044b3f8c820d5da6405ccdec17ecebd9638267R3397))
*  Update the `llama.cpp` file to use the new API and pass the number 1 to the `ggml_metal_init` function ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR343-R345), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL761), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL2648-R2674), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR2831-R2833), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL2969-R3000), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL3123), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL3143-R3171), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL3229), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efL3249-R3276))
*  Enable the `test-grad0.c` test and update it to use the new API and ignore the double promotion warning ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-4461c617ceae0c7e0206622aacdf555aedd62eb129c2efb74c84fa1567bcbe0dL13-R13), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-719f57ae1dee9d562185e8fbe8a3c57b4b346c3a09c0fab71af3a7a5bff66cbdR13-R14), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-719f57ae1dee9d562185e8fbe8a3c57b4b346c3a09c0fab71af3a7a5bff66cbdL52-R54), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-719f57ae1dee9d562185e8fbe8a3c57b4b346c3a09c0fab71af3a7a5bff66cbdL162-R171), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-719f57ae1dee9d562185e8fbe8a3c57b4b346c3a09c0fab71af3a7a5bff66cbdL218-R230), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-719f57ae1dee9d562185e8fbe8a3c57b4b346c3a09c0fab71af3a7a5bff66cbdL239-R251), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-719f57ae1dee9d562185e8fbe8a3c57b4b346c3a09c0fab71af3a7a5bff66cbdL255-R265))
*  Update the `test-opt.c` test to use the new API and ignore the double promotion warning ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-b0455748dd7cc28380469ad1ae439aac0dcb65d0bc8e9712fdcafd6adbda408fR10), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-b0455748dd7cc28380469ad1ae439aac0dcb65d0bc8e9712fdcafd6adbda408fL36-R37), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-b0455748dd7cc28380469ad1ae439aac0dcb65d0bc8e9712fdcafd6adbda408fL117-R118), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-b0455748dd7cc28380469ad1ae439aac0dcb65d0bc8e9712fdcafd6adbda408fL140-R145), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-b0455748dd7cc28380469ad1ae439aac0dcb65d0bc8e9712fdcafd6adbda408fL151-R156))
*  Add a timeout option to the `ctest` command for the GitHub workflow jobs to avoid the tests from hanging or running too long and add a new environment variable `GGML_NLOOP` to control the number of iterations for the benchmark tests ([link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721L19-R21), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721L67-R69), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721L102-R104), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721L150-R156), [link](https://github.com/ggerganov/llama.cpp/pull/1999/files?diff=unified&w=0#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721L252-R255))

### CODE REVIEW

**Comment by slaren** [comment]
>ggml_graph_compute_plan() MUST be called because it also sets node->n_tasks. The work_size depends on n_tasks.

I think that `n_tasks` should be removed from `ggml_tensor`. For now, the easiest way to address may be to move it to the new `ggml_cgraph_context`, but eventually I think that this is something that `ggml_graph_compute` should be able to determine automatically when executing a node.

Good job!

**Comment by slaren** [comment]
>Of course, `n_tasks` should belong to the compute facility I think, it's ideal to migrate to some place else.

Yes precisely! That's what I was thinking as well.

I am not sure that we need a new struct for this, is there any other data from the tensor would belong there? I am thinking that adding a simple `int n_tasks[GGML_MAX_NODES]` to `ggml_cgraph_context` with a 1:1 mapping between the graph nodes should do it. It is not great, but for now it will solve the problem, and eventually I think that we can remove it and have `ggml_graph_compute` determine this automatically instead.

**Comment by mqy** [comment]
> is there any other data from the tensor would belong there?

No way.

> adding a simple `int n_tasks[GGML_MAX_NODES]`

Great! The n_tasks array is good enough.

> and eventually I think that we can remove it and have `ggml_graph_compute` determine this automatically instead.

That's will be a  long story. I had tried add  another enum to `ggml_task_type` named `GGML_TASK_CONFIG`, `ggml_graph_compute()` runs the planning before computing by calling `ggml_graph_forward()` for every node, similar to the planning idea in this PR.  At that time I prefer this kind of design because actually the computing codes should know exactly what they can provide (parallel?) and want are required (wsize?). That's easy to maintain-- just have a glance at the actual codes, less chances to mis-config. 

I think `ggml_graph_compute()` MAY NOT necessary to  hard code everything in the long run, it's better to build some kinds of new protocol/contract between the scheduler and actual computing codes. To achieve this, we can construct another slightly abstract data structure of top of {node, params}, with `methods` such as `config()`, `compute()`.  With this design, we could create and register several compute vendors (with id/name for example) for every kind of node,  this provide great flexibility for evaluating experimental algorithms, benchmarking. With the compute vendor id, we are hopefully able to define a compute plan offline (if this is unnecessary, at least we can specify certain vendors) , this is similar to the dumped cgraph.

Just some random thoughts :D

**Comment by slaren** [comment]
That's very similar to what I have been thinking. I am working on a CUDA implementation that can execute `ggml_cgraphs` directly, and what it needs to do that is very similar to this, a method to "prepare" or "plan" the graph, and another to execute the "prepared" graph. Eventually, we will want a common interface to all compute backends so that users can use any of them interchangeably, without having to add specific code for each one.

While doing this, I think that we should also remove `n_threads` from `ggml_cgraph`, since this parameter only relevant to the CPU compute backend and not applicable to other compute backends.

**Comment by mqy** [comment]
> Eventually, we will want a common interface to all compute backends so that users can use any of them interchangeably, without having to add specific code for each one.

Yes, I saw the changes during the last two months, getting better and better. Both CUDA and CL implementations impose potential de-coupling requirements.

> While doing this, I think that we should also remove `n_threads` from `ggml_cgraph`, since this parameter is only relevant to the CPU compute backend and not applicable to the other compute backends.

Sure, similar to `n_tasks` in tensor as well, but no idea where to put it when compute, hello ggml_cgraph_context?

At present, I'm not sure but looks like for cross cgraph computing, the `n_threads` is still required, especially when a node runner is allowed to hijack the computing flow or fallback to default vendor -- the CUDA way. In cross cgraph context, `n_threads` or thread_ctx should be owned by something like `compute session`? 

I think it's OK to leave `n_threads` there for a while, GPU codes can simply ignore it, unless the GPU implementation has it's complete cgraph that is not shared with CPU, this is yet another long story.

**Comment by howard0su** [comment]
I like the idea. However, I am not favoring the implementation:
1. We should not rely on consumer's behavior says don't touch plan. We should not expose cgraph_compute_context at all or maintain the state elsewhere.
2. a common API pattern for external provided buffer is like: Pass NULL as buf, and buffer size into API. API will return the size it expected but don't do the real computation. Pass a valid pointer but the size is less than expected, the function will fail in the same way.
3. plan is not a clear to indicate this function is needed to be called before compute API.  compute_v2 takes a result of this plan will make the API explict. also maybe create compute_context is better API name. plan is too general.

**Comment by mqy** [comment]
> I am not favoring the implementation

Yes, totally make points. A general context is used for future extensibility, but should be seen as over design. If you have read previous comments you would get the conclusion that a general context tends be "extended" as a basket.

So, if we can't tell what we want beyond current requirement, a specialized API should be the best choice.
Anyway, I don't think plan is a bad idea, the actual problem is we haven't decouple the calculating wsize from setting n_tasks. Perhaps we'd reconsider this design after splitting that part of codes.

**Comment by ggerganov** [comment]
> I think that n_tasks should be removed from ggml_tensor

Agree

Good points by @howard0su 

Overall, I think this is on the right track and we should finalize the PR and merge it.

There is no need to worry about backwards compatibility (i.e. the "_v2" stuff is not necessary). Just assume that we have just started developing the API and nobody has been using it and expect it to be compatible. We will start to worry about this kind of things after the v1.0 release sometime in the future.

**Comment by mqy** [comment]
> There is no need to worry about backwards compatibility

This is great, I'll rewrite this PR, show you later.

File: ggml.c [hunk 1]
```diff
@@ -16856,23 +16856,20 @@ static thread_ret_t ggml_graph_compute_thread(void * data) {
     return 0;
 }
 
-void ggml_graph_compute(struct ggml_context * ctx, struct ggml_cgraph * cgraph) {
-    const int n_threads = cgraph->n_threads;
+// Prepare for graph computing.
+// Will set: node->n_tasks, ctx->{work_size, planned}
+void ggml_graph_compute_plan(struct ggml_cgraph_context * ctx, struct ggml_cgraph * cgraph) {
```
Comment (mqy) [inline ggml.c:16861]: I saw https://github.com/ggerganov/ggml/issues/299#issuecomment-1616698106
Yes, it's better to return `ggml_cgraph_context` as @slaren suggested. 
```
ggml_cgraph_context plan = ggml_graph_compute_plan(&gf);
```

File: tests/test-grad0.c [hunk 2]
```diff
@@ -1154,7 +1208,7 @@ int main(int argc, const char ** argv) {
                             continue;
                         }
 
-                        struct ggml_tensor * f = ggml_sum(ctx0, ggml_rope(ctx0, x[0], n_past, n_rot, mode));
+                        struct ggml_tensor * f = ggml_sum(ctx0, ggml_rope(ctx0, x[0], n_past, n_rot, mode, 0));
```
Comment (mqy) [inline tests/test-grad0.c:1211]: Is this right?

File: tests/test-grad0.c [hunk 3]
```diff
@@ -215,15 +215,35 @@ bool check_gradient(
     }
 
     struct ggml_cgraph gf = ggml_build_forward (f);
-    gf.n_threads = n_threads;
 
     struct ggml_cgraph gb = ggml_build_backward(ctx0, &gf, false);
```
Comment (mqy) [inline tests/test-grad0.c:219]: Sadly, ggml_build_backward was broken, I'll dig into it.

Comment (mqy) [reply tests/test-grad0.c:219]: Clean build by deleting `build` dir  resolved this issue, no idea what's wrong. The output is similar to master branch.

```
$ time ./bin/test-grad0 
test-grad0: iter:0/4
GGML_N_THREADS = 4
test-grad0: iter:1/4
test-grad0: iter:2/4
cross_entropy_loss: ndims=3, i=0, k=0, x0=-0.951187, xm=-1.051187, xp=-0.851187, f0=11.762405, f1=11.762405, g0=0.000000, g1=-0.571868, eps=0.100000, error_abs=0.571868, error_rel=0.000000
test-grad0: iter:3/4
```

File: tests/test-grad0.c (hunk 2 above)
Comment (mqy) [reply tests/test-grad0.c:1210]: Looks like it works, because:
- the test runs without error;
- and I saw `n_ctx` is set as `0` in other places.

**Comment by mqy** [comment]
The PR description was updated. You may want to have a look at it. Apart from `main` and `perplexity`, also tested:
- tests/test-grad0
- tests/test-opt
- examples/benchmark
- examples/baby-llama
- examples/train-text-from-scratch (force stop at Example 2, opt iter 29)

No crash, all of them output reasonable text.

So this PR is ready to review again.

**Comment by slaren** [comment]
Looks good, I only have a few minor nits:
- In llama.cpp, to avoid allocations in every eval, the work buffer memory could be stored as a `std::vector` in `lama_context`. Just `resize` it to the `work_size`, and if it is already big enough, it won't reallocate memory.
- More generally, in all the C++ code, I would suggest always using `std::vector` instead of `malloc`/`free`
- I am not sure about `ggml_graph_compute_sugar`

**Comment by mqy** [comment]
https://github.com/ggerganov/llama.cpp/pull/1999/commits/bf63002af94bc8fe39bca1a44e1344fad420f96b is the latest commit per @slaren 's suggestion. 

TODO: `The sugar is used by baby-llama. It calls ggml_graph_compute again and again. I suggest we delay it to next PR.`

Rebased on to master.

**Comment by mqy** [comment]
rebased

**Comment by ggerganov** [comment]
I've refactored the changes:

- add `ggml_graph_compute_with_ctx()` - similar to the old way of graph computation
- rename `struct ggml_graph_compute_plan` to `struct ggml_cplan`
- rename `ggml_graph_compute_make_plan()` to `ggml_graph_plan()`
- factored common "plan + compute" code in a helper function `ggml_graph_compute_helper()`
- enabled `test-grad0` test
- fixed Metal build

Overall the change is good.
A positive side-effect is that the user can now control the number of tasks for each op. This can be utilized also when creating custom operators, if we expose the `struct ggml_compute_params` to the custom function signatures.

@mqy @slaren Please take a look

**Review by ggerganov — APPROVED** [review]

**Comment by monatis** [comment]
> rename ggml_graph_compute_make_plan() to ggml_graph_plan()

I would suggest `ggml_cplan_make()` --both short as intended and also consistent with the struct naming.

**Comment by slaren** [comment]
I think this looks good.

>A positive side-effect is that the user can now control the number of tasks for each op. This can be utilized also when creating custom operators, if we expose the struct ggml_compute_params to the custom function signatures.

Eventually we should support custom ops on the GPU backends too. It will be slow since it will require copying data back and forth from the device, but that's still a lot better than not supporting them at all. So whenever we do this, I think we should do it in a backend-agnostic way, rather than tying it to the CPU backend.

**Comment by mqy** [comment]
> I've refactored the changes:

Looks good. Verified main and test-grad0.

> A positive side-effect is that the user can now control the number of tasks for each op.

- Increasing n_tasks[i] after plan may cause work buffer overflow.
- For OPs that can only be run with n_tasks == 1,  it's UB if run with n_tasks > 1.

Anyway, better than never.

**Comment by ggerganov** [comment]
> I would suggest ggml_cplan_make() --both short as intended and also consistent with the struct naming.

Good suggestion.
I like grouping names with common prefixes and currently `ggml_graph_plan()` fits neatly with the rest: `ggml_graph_compute()`, `ggml_graph_reset()` so I prefer not to isolate it in a new prefix. If we start doing more stuff with the plans, like `ggml_cplan_optimize()`, `ggml_cplan_print()`, etc, we can rename it.

Btw, this reminds me that `ggml_graph_xxx` probably should have been `ggml_cgraph_xxx`. Shall we change it?

> Increasing n_tasks[i] after plan may cause work buffer overflow.

Ah good point. So the developer has to be careful and modify `n_tasks` only for their own custom operators or be familiar with the inner workings of the `ggml` ops.

**Pull request merged**